### PR TITLE
Fix view change stuck issue

### PR DESCRIPTION
**Pull request by rlan35**
View change still got stuck when manually abort consensus on commit phase every 7 block, with following code added in leader.go:

```
 func (consensus *Consensus) onCommit(msg *msg_pb.Message) {
+       utils.Logger().Info().Msgf("ViewChanging %d %d", consensus.GetCurBlockViewID(), consensus.GetViewChangingID())
+       if consensus.GetCurBlockViewID()%7 == 0 {
+               return
+       }
        recvMsg, err := consensus.ParseFBFTMessage(msg)
        if err != nil {
                consensus.getLogger().Debug().Err(err).Msg("[OnCommit] Parse pbft message failed")
```

There are two bugs fixed in this PR:

1. When selfCommit, we need to use AddNewVote, instead of SubmitVote. SubmitVote is a lower level logic which doesn't take care of voting power calculation. Fixed and also refactored to make submitVote private.

2. The view changing id changes in cycle, like 7, 8, 9, 10, 7, 8, 9, 10, 7.....  rather than monotonically increase. Actually View ID should never decrease.

Also made view changing id to grow with block view id to make things cleaner and more consistent. Basically, view change id always >= block view id.

PS: also find a corner case leaderPubKey mistake. See the todo added. The mitigation is to make block sync happen less frequent. Because block sync will make the node reject all consensus messages, which is not good for stability.

### CODE REVIEW

File: api/service/syncing/syncing.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ const (
 	TimesToFail                     = 5 // downloadBlocks service retry limit
 	RegistrationNumber              = 3
 	SyncingPortDifference           = 3000
-	inSyncThreshold                 = 0    // when peerBlockHeight - myBlockHeight <= inSyncThreshold, it's ready to join consensus
+	inSyncThreshold                 = 1    // when peerBlockHeight - myBlockHeight <= inSyncThreshold, it's ready to join consensus
```
Comment (LeoHChen) [inline api/service/syncing/syncing.go:35]: why this change?

File: consensus/quorum/quorum.go [hunk 2]
```diff
@@ -273,7 +273,7 @@ func (s *cIdentities) SignersCount(p Phase) int64 {
 	}
 }
 
-func (s *cIdentities) SubmitVote(
+func (s *cIdentities) submitVote(
```
Comment (LeoHChen) [inline consensus/quorum/quorum.go:276]: need a better description on how to use this function, can you add some comments.

File: consensus/view_change.go [hunk 3]
```diff
@@ -131,9 +131,8 @@ func (consensus *Consensus) getNextViewID() (uint64, time.Duration) {
 	if curTimestamp <= blockTimestamp {
 		return consensus.fallbackNextViewID()
 	}
-	totalNode := consensus.Decider.ParticipantsCount()
 	// diff is at least 1, and it won't exceed the totalNode
-	diff := uint64(((curTimestamp - blockTimestamp) / viewChangeTimeout) % int64(totalNode))
+	diff := uint64((curTimestamp - blockTimestamp) / viewChangeTimeout)
```
Comment (LeoHChen) [inline consensus/view_change.go:135]: this is good, we already take care of nextLeader rotation with mod.

**Review by LeoHChen — APPROVED** [review]
some minor comments, LGTM.  Can you add some comments in AddNewVote function, so that we won't misuse it later.

File: api/service/syncing/syncing.go (hunk 1 above)
Comment (rlan35) [reply api/service/syncing/syncing.go:35]: see the todo added. because block sync will make the validator fail to accept consensus message. and in the rare case when the sync check happens in a bad timing, the validator could enter sync mode prematurely and miss consensus messages.

File: consensus/quorum/quorum.go (hunk 2 above)
Comment (rlan35) [reply consensus/quorum/quorum.go:276]: Done.

File: consensus/quorum/quorum.go [hunk 4]
```diff
@@ -81,7 +81,8 @@ type ParticipantTracker interface {
 // SignatoryTracker ..
 type SignatoryTracker interface {
 	ParticipantTracker
-	SubmitVote(
+	// This func shouldn't be called directly from outside of quorum. Use AddNewVote instead.
+	submitVote(
```
Comment (JackyWYX) [inline consensus/quorum/quorum.go:85]: Consider removing from interface?

File: consensus/validator.go [hunk 5]
```diff
@@ -272,6 +270,11 @@ func (consensus *Consensus) onCommitted(msg *msg_pb.Message) {
 
 	consensus.FBFTLog.AddMessage(recvMsg)
 
+	if recvMsg.BlockNum > consensus.blockNum {
```
Comment (JackyWYX) [inline consensus/validator.go:273]: If a prepared message is missing, onCommitted will never spin state sync since the it is already returned in 

```
	blockObj := consensus.FBFTLog.GetBlockByHash(recvMsg.BlockHash)
	if blockObj == nil {
		consensus.getLogger().Debug().
			Uint64("blockNum", recvMsg.BlockNum).
			Uint64("viewID", recvMsg.ViewID).
			Str("blockHash", recvMsg.BlockHash.Hex()).
			Msg("[OnCommitted] Failed finding a matching block for committed message")
		return
	}
```

Is this what we expect?

Comment (rlan35) [reply consensus/validator.go:273]: yes, we want to spin up the sync when it's very certain. Don't want to turn to sync mode prematurely. It's risky.

Comment (JackyWYX) [reply consensus/validator.go:273]: I mean this sync code is unreachable since the block from prepared message is missing.

File: consensus/quorum/quorum.go (hunk 4 above)
Comment (rlan35) [reply consensus/quorum/quorum.go:85]: it's used as typing for different type of structs. So not easy to remove.

**Comment by rlan35** [comment]
jenkins, test this please.

File: consensus/validator.go (hunk 5 above)
Comment (rlan35) [reply consensus/validator.go:273]: Ok, then we probably don't need to have this sync code in oncommitted. What do you think?

Comment (JackyWYX) [reply consensus/validator.go:273]: OK, it's actually a trade off between catch up performance and how sync mode effects the rest of validator consensus logic. I definitely agree with reducing complexity of the consensus logic for now, and refactoring the message handling logic for sync mode can be a good todo in the future.

**Review by JackyWYX — APPROVED** [review]
